### PR TITLE
feat: file filter considers tracked files before applying ignores

### PR DESCRIPTION
**Pull request by CatalinSnyk**
### Description

TLDR: The file filter now considers the tracked file state before applying .gitignore rules. Meaning, a tracked file will be included even if it's matched by a rule. 

### Checklist

- [x] Tests added and all succeed (`make test`)
- [x] Regenerated mocks, etc. (`make generate`)
- [x] Linted (`make lint`)
- [x] Test your changes work for the CLI (local tested) :loading: CLI PR

---

> [!NOTE]
> **Medium Risk**
> Changes which files are included in scans when the flag is enabled, with git index reads on the filtering path; behavior is feature-flagged and degrades to legacy matching on git errors.
> 
> **Overview**
> Adds **tracked-file-aware `.gitignore` handling** behind `internal_snyk_gitignore_respect_tracked_files_enabled` (CLI-1411), wired to rollout flag `clientFileFilterGitignore_TrackedFilesRollout`.
> 
> When the flag is on, `.gitignore` globs are tagged with a `#gitignore:` prefix and `GetFilteredFiles` uses a git-index-backed predicate: files **in the index** that match `.gitignore` stay in the scan set (Git-style), while **`.dcignore` and `.snyk` exclusions still drop tracked paths**. Untracked paths matching `.gitignore` are excluded as before.
> 
> The implementation opens the repo via **go-git**, resolves the scan root against the worktree (subdirectory scans and symlinked parents), and falls back to the previous glob-only matcher if the repo cannot be opened.
> 
> Adds broad regression tests for staged/committed tracked files, nested rules, negation, and index edge cases, plus an optional repo benchmark comparing flag on vs off.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c703bd66714643aabc30defa4b7493eb5d028922. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: pkg/utils/file_filter.go [hunk 1]
```diff
@@ -233,8 +236,16 @@ func (fw *FileFilter) GetRules(ruleFiles []string) ([]string, error) {
 func (fw *FileFilter) GetFilteredFiles(filesCh chan string, globs []string) chan string {
 	var filteredFilesCh = make(chan string)
 
-	// create pattern matcher used to match filesToFilter to glob patterns
-	globPatternMatcher := gitignore.CompileIgnoreLines(globs...)
+	var isFileExcluded func(string) bool
+	if fw.config.GetBool(FF_GITIGNORE_RESPECT_TRACKED_FILES) {
+		isFileExcluded = fw.trackedFileExclusionPredicate(globs)
+	} else {
```
Comment (danskmt) [inline pkg/utils/file_filter.go:242]: @PeterSchafer @CatalinSnyk the strategy of prepending a character to .gitignore rules creates a question that must be answered here:
-> How do we guarantee the list of **globs** contains the prefixes attached to it? Consumers may call with any kind of list, they can add, remove, modify the way they want before calling **GetFilteredFiles**, so we are not ensuring the behavior will be applied as we want. For example, what if the consumers do not use the "GetRules" method to build the globs, and build their own?

File: pkg/utils/file_filter.go
Comment (danskmt) [inline pkg/utils/file_filter.go:233]: @PeterSchafer @CatalinSnyk Please note that prepending a character to .gitignore rules is changing the behavior of **GetRules** function. Consumers may expect the rules as they are, with no modifications. However, **buildGlobs** at line 227 is now modifying such rules. The question is:
-> Do we need to properly communicate this change of behavior of **GetRules**? It will not be honest in the sense of giving the raw rules, but will be giving the modified rules instead. Also, is there any impact if the consumer wants to add their own rules to globs list?

File: pkg/utils/file_filter.go (hunk 1 above)
Comment (danskmt) [reply pkg/utils/file_filter.go:242]: e.g. **cli-extension-iac** builds its own globs and never calls **GetRules**

File: pkg/utils/file_filter.go
Comment (danskmt) [reply pkg/utils/file_filter.go:233]: Someone relying on **GetRules** for some other logic will take the gitignore rules as comments, this is a risk for consumers

File: pkg/utils/file_filter.go [hunk 2]
```diff
@@ -287,13 +299,115 @@ func (fw *FileFilter) buildGlobs(ignoreFiles []string) ([]string, error) {
...
+		if err != nil {
+			return allMatcher.MatchesPath(normalizedPath)
+		}
+
+		if trackedFilesToPreserve[filepath.ToSlash(relativePath)] {
+			return otherMatcher.MatchesPath(normalizedPath)
+		}
+
+		return allMatcher.MatchesPath(normalizedPath)
+	}
```
Comment (danskmt) [inline pkg/utils/file_filter.go:384]: What about:
```go
	trackedPaths := make(map[string]struct{}, len(gitIndex.Entries))
	for _, entry := range gitIndex.Entries {
		trackedPaths[entry.Name] = struct{}{}
	}

	return func(file string) bool {
		normalizedPath := filepath.ToSlash(file)
		relativePath, err := filepath.Rel(repoRoot, file)
		if err != nil {
			return allMatcher.MatchesPath(normalizedPath)
		}

		if _, tracked := trackedPaths[filepath.ToSlash(relativePath)]; tracked {
			if gitignoreMatcher.MatchesPath(normalizedPath) {
				return otherMatcher.MatchesPath(normalizedPath)
			}
		}

		return allMatcher.MatchesPath(normalizedPath)
	}
```

Comment (CatalinSnyk) [reply pkg/utils/file_filter.go:384]:
```go
if _, tracked := trackedPaths[filepath.ToSlash(relativePath)]; tracked {
    if gitignoreMatcher.MatchesPath(normalizedPath) {
        return otherMatcher.MatchesPath(normalizedPath)
    }
}
```
This more/less does the same thing as #685 (almost, since there we used a slice of sorted paths & binary search). But this will store all the tracked files, and not the intersection of tracked & .gitignored files in the map. On bigger repos this might take up more memory.

Comment (danskmt) [reply pkg/utils/file_filter.go:384]: I ran a benchmarking with claude and it seems it improves speed by quite a lot:
image

image

Comment (CatalinSnyk) [reply pkg/utils/file_filter.go:384]: i would say this should & will be investigated further in the follow-up ticket. The benchmark numbers do look great, but not sure how accurate they are given the entries kept column is yielding quite different results 😄

File: pkg/utils/file_filter.go (hunk 1 above)
Comment (basti-snyk) [reply pkg/utils/file_filter.go:242]: That's ok, we do not need to cater to people NOT using a function that was exported.

File: pkg/utils/file_filter.go
Comment (basti-snyk) [reply pkg/utils/file_filter.go:233]: This is theoretical. Currently, external users are osflows, secrets and snyk-ls. They all want this change.

**Review by danskmt — APPROVED** [review]
